hal/src/thumbv7em/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1371 @@
+    /// See datasheet for more information.
+    pub fn irda_encoding(mut self, pulse_length: Option<u8>) -> Self {
+        self._irda_encoding(pulse_length);
+        self
+    }
+}
+
+impl<P, C> Config<P, C>
+where
+    P: GetCapability + ValidPads,
+    C: CharSize,
+    Self: ValidConfig,
+{
+    /// Enable the UART peripheral and return a [`Uart`] struct.
+    ///
+    /// UART transactions are not possible until the peripheral is enabled.
+    /// This method is limited to [`ValidConfig`]s
+    #[inline]
+    pub fn enable(mut self) -> Uart<Self, P::Capability> {
+        self._enable();
+        Uart {
+            config: self,
+            capability: PhantomData,
+        }
+    }
+}
+
+//=============================================================================
+// AnyConfig
+//=============================================================================
+
+/// Type class for all possible [`Config`] types
+///
+/// This trait uses the [`AnyKind`] trait pattern to create a [type class] for
+/// [`Config`] types. See the `AnyKind` documentation for more details on the
+/// pattern.
+///
+/// In addition to the normal, `AnyKind` associated types. This trait also
+/// copies the [`Sercom`] and [`Word`] types, to make it easier to apply
+/// bounds to these types at the next level of abstraction.
+///
+/// [`AnyKind`]: crate::typelevel#anykind-trait-patter
+pub trait AnyConfig: Sealed + Is<Type = SpecificConfig<Self>> {
+    type Sercom: Sercom;
+    type Pads: ValidPads<Sercom = Self::Sercom>;
+    type Word: 'static;
+    type CharSize: CharSize<Word = Self::Word>;
+}
+
+/// Type alias to recover the specific [`Config`] type from an implementation of
+/// [`AnyConfig`]
+pub type SpecificConfig<C> = Config<<C as AnyConfig>::Pads, <C as AnyConfig>::CharSize>;
+
+/// Type alias to recover the specific [`Sercom`] type from an implementation of
+/// [`AnyConfig`]
+pub type ConfigSercom<C> = <C as AnyConfig>::Sercom;
+
+impl<P, C> AsRef<Self> for Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    #[inline]
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<P, C> AsMut<Self> for Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    #[inline]
+    fn as_mut(&mut self) -> &mut Self {
+        self
+    }
+}
+
+impl<P, C> Sealed for Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+}
+
+impl<P, C> AnyConfig for Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    type Sercom = P::Sercom;
+    type Word = C::Word;
+    type Pads = P;
+    type CharSize = C;
+}
+
+//=============================================================================
+// ValidConfig
+//=============================================================================
+
+/// Marker trait for valid UART [`Config`]urations
+///
+/// A functional UART peripheral must have, at a minimum either a Rx or a Tx
+/// [`Pad`].
+pub trait ValidConfig: AnyConfig {}
```

Comment:
If you move `Capability` to an associated type of `ValidPads`, I don't think you'll need separate implementations here.